### ISSUE TITLE
Uninformative error on self-conflicting types

### ISSUE THREAD

**Issue by tormeh** [oxidecomputer/typify#485]
I used the following OpenAPI schema:
```json
{
  "components": {
    "schemas": {
      "Foo": {
        "allOf": [
          {
            "$ref": "#/components/schemas/Bar"
          },
          {
            "properties": {
              "value": {
                "type": "string"
              }
            }
          }
        ],
        "required": [
          "value"
        ]
      },
      "Bar": {
        "properties": {
          "value": {
            "type": "object"
          }
        }
      }
    }
  },
  "info": {
    "title": "",
    "version": "1.0"
  },
  "openapi": "3.0.1",
  "paths": {
  }
}
```

I expected an error saying that the type of `value` in `Foo` was ambiguous. I got:
```
thread 'main' panicked at /Users/[my_user]/.cargo/registry/src/index.crates.io-6f17d22bba15001f/typify-impl-0.0.15/src/merge.rs:265:70:
called `Result::unwrap()` on an `Err` value: ()
```

It seems silly in this example, but the original schema was 10000 lines long, so it was quite tedious to isolate the issue.

## Meta

Progenitor version used was 0.5.0.
I used a build.rs file to invoke progenitor.

**Comment by ahl** [comment]
I've moved this to the typify repo since the code it pertains to lives here. Progenitor hands of processing of types to typify.

First, yes, the error message need to be improved. There's ongoing work to maintain context so that, in a case such as this one, the error pointed you to the specific part of the large specification document where the problem was encountered.

It seems like the schema for Bar is wrong for its intended use. Perhaps it should be something like:

```json
      "Bar": {
        "properties": {
          "value": {}
        }
      }
```

or (equivalently):

```json
      "Bar": {
        "properties": {
          "value": true
        }
      }
```

I'm not sure what weight to put on the fact that Bar isn't used directly. In general, typify (and progenitor) studiously avoid heuristic handling of potentially misconstructed schemas. Specifically, when a schema is innately unsatisfiable, there is no intrinsically correct answer regarding which constraints one might relax to resolve that unsatisfiability. In fact, the direction we've been going is to use "never" types in situations such as these (`enum Never {}` i.e. an enum with no variants that is therefore uninstantiatable).

My general recommendation has been to apply a JSON patch to the input document before passing it to progenitor (or to typify).

**Comment by tormeh** [comment]
I agree. My solution for now is to preprocess the schema in build.rs, as it is the schema that is wrong.

I don't think a lot of context is necessary, although I wouldn't exactly complain about a rustc-style error message with line numbers and everything. It might be enough to state the name of the type that couldn't be determined. Just knowing that `Foo.value` was the issue would have helped a lot, though the error would ideally say `Foo` and `Bar` conflicted with regards to `Foo.value`.

**Comment by ahl** [comment]
> I don't think a lot of context is necessary, although I wouldn't exactly complain about a rustc-style error message with line numbers and everything. It might be enough to state the name of the type that couldn't be determined. Just knowing that `Foo.value` was the issue would have helped a lot, though the error would ideally say `Foo` and `Bar` conflicted with regards to `Foo.value`.

Absolutely: I really just want to give you a json path since some specs don't contain newlines so saying "somewhere on line 1" wouldn't be helpful.
